hook-emulator/c_wrappers/main.c

```diff
@@ -22,133 +51,508 @@ enum MESSAGE_TYPES
-// Keylet response - <issuer(20)><currency(3)><balance(8)XFL><limit(8)XFL>
+// Message response format - <RETURN CODE(1)><DATA>
+// Emit response - <return code(1)><tx hash(32)>
+// Trustline response - <return code(1)><balance(8)XFL><limit(8)XFL>
+// State get response - <return code(1)><value(128)>
+// State set response - <return code(1)>
 // '''''''''''''''''''''''''''''''''''''''''''''''''''''
 
 // ' C_WRAPPER --> JS (Write to STDOUT from C_WRAPPER) '
-// Trace - <TYPE:TRACE(1)><trace message>
-// Transaction emit - <TYPE:EMIT(1)><account(20)><1 for xrp and 0 for iou(1)><[XRP: amount in buf(8)XFL][IOU: <issuer(20)><currency(3)><amount in buf(8)XFL>]><destination(20)><memo count(1)><[<TypeLen(1)><MemoType(20)><FormatLen(1)><MemoFormat(20)><DataLen(1)><MemoData(128)>]><ledger_hash(32)><ledger_index(8)>
-// Keylet request - <TYPE:KEYLET(1)><issuer(20)><currency(3)>
+// Message request format - <TYPE(1)><DATA>
+// Trace request - <type:TRACE(1)><trace message>
+// Transaction emit request - <type:EMIT(1)><prepared transaction buf>
+// Trustline request - <type:TRUSTLINE(1)><address(20)><issuer(20)><currency(3)>
+// State set request - <type:STATE_GET(1)><key(32)><value(128)>
+// State get request - <type:STATE_SET(1)><key(32)>
 // '''''''''''''''''''''''''''''''''''''''''''''''''''''
 // -----------------------------------------------------
 
-void write_stdout(const uint8_t *buf, const int len)
+/**
+ * Write given buffer to the STDOUT.
+ * Adds the data length as a prefix according to the message protocol.
+ * @param write_buf Buffer to be sent.
+ * @param write_len Length of the buffer.
+ * @return -1 if error, otherwise write length.
+*/
+int write_stdout(const uint8_t *write_buf, const int write_len)
 {
-    const int outlen = 4 + len;
+    const int outlen = 4 + write_len;
     char out[outlen];
-    out[0] = len >> 24;
-    out[1] = len >> 16;
-    out[2] = len >> 8;
-    out[3] = len;
-    for (int i = 0; i < len; i++)
-        out[i + 4] = buf[i];
-
-    write(STDOUT_FILENO, out, outlen);
-    fflush(stdout);
+
+    // Populate the data length as prefix.
+    out[0] = write_len >> 24;
+    out[1] = write_len >> 16;
+    out[2] = write_len >> 8;
+    out[3] = write_len;
+
+    // Populate the data.
+    memcpy(&out[4], write_buf, write_len);
+
+    return (write(STDOUT_FILENO, out, outlen) == -1 || fflush(stdout) == -1) ? -1 : write_len;
+}
+
+/**
+ * Read data from STDIN to the buffer.
+ * @param read_buf Buffer to be read.
+ * @return -1 if error, otherwise read length.
+*/
+int read_stdin(uint8_t **read_buf)
+{
+    // Read the response from STDIN.
+    // Data length is added from the protocol in first 4 bytes, so we skip those.
+    uint8_t buf[MAX_READ_LEN];
+
+    if (read(STDIN_FILENO, buf, sizeof(buf)) == -1)
+        return -1;
+
+    // Read the data length.
+    uint32_t len = buf[3] | (buf[2] << 8) | (buf[1] << 16) | (buf[0] << 24);
+
+    // Populate the data to buffer.
+    *read_buf = (uint8_t *)malloc(len);
```

Comment:
Instead of "malloc"ing inside the function, better to let the caller pre-allocate memory in advance. Basically design the read_stdin like the read() system call. Specify the max length the caller expect to read. I think in our case we know stdin data will always be less than some boundry (say 128KB). Then the caller can allocate memory for max limit and the read_stdin will return the number of actual bytes read.